src/hooks/usePendingSwapData.ts

```diff
@@ -0,0 +1,214 @@
+        const newItemId = event.args?.itemId as BigNumber | null
+        if (newItemId == null) return
+        void event.getBlock().then((block) => {
+          void fetchPendingSwapInfo(
+            bridgeContract,
+            newItemId,
+            block.timestamp,
+            chainId,
+          ).then((fetchedPendingSwap) => {
+            if (fetchedPendingSwap == null) return
+            setPendingSwaps((existingState) => {
+              return [
+                fetchedPendingSwap,
+                ...existingState.filter(({ itemId }) => itemId !== newItemId),
+              ]
+            })
+          })
+        })
+      }
+
+      VIRTUAL_SWAP_TOPICS.forEach((topic) => {
+        void bridgeContract.on(topic, eventListener)
+      })
+      return () => {
+        VIRTUAL_SWAP_TOPICS.forEach((topic) => {
+          bridgeContract.off(topic, eventListener)
+        })
+      }
+    }
+    void attachPendingSwapEventListeners()
+    void fetchExistingPendingSwaps()
+  }, [account, library, chainId, bridgeContract, queryStartBlock])
+  return pendingSwaps
+}
+
+/**
+ * Create filters for each PendingSwap type
+ */
+function buildEventFilters(bridgeContract: Bridge, account: string) {
+  return VIRTUAL_SWAP_TOPICS.map((topic) => {
+    return bridgeContract.filters[topic](account, null, null, null, null, null)
+  })
+}
+
+enum BridgePendingSwapTypes {
+  Null,
+  TokenToSynth,
+  SynthToToken,
+  TokenToToken,
+}
+async function fetchPendingSwapInfo(
+  bridgeContract: Bridge,
+  itemId: BigNumber,
+  timestampInSeconds: number, // in seconds
+  chainId: ChainId,
+) {
+  let result = null
+  try {
+    // this will throw if the itemId has already fully resolved
+    // afik we don't have a better way of fetching current pendingSwaps
+    // DEVELOPER: for this reason we don't use multicall
+    const pendingSwapInfo = await bridgeContract.getPendingSwapInfo(itemId)
+    const pendingSwapType = pendingSwapInfo.swapType
+    let swapType
+    if (pendingSwapType === BridgePendingSwapTypes.TokenToSynth) {
+      swapType = SWAP_TYPES.TOKEN_TO_SYNTH
+    } else if (pendingSwapType === BridgePendingSwapTypes.SynthToToken) {
+      swapType = SWAP_TYPES.SYNTH_TO_TOKEN
+    } else if (pendingSwapType === BridgePendingSwapTypes.TokenToToken) {
+      swapType = SWAP_TYPES.TOKEN_TO_TOKEN
+    } else {
+      swapType = SWAP_TYPES.INVALID
+    }
+    const synthTokenFrom = getTokenByAddress(
+      pendingSwapInfo.synth,
+      chainId,
+    ) as Token
+    const tokenTo = getTokenByAddress(pendingSwapInfo.tokenTo, chainId) as Token
+    const settleableAtTimestamp = new Date(
+      (timestampInSeconds + parseInt(pendingSwapInfo.secsLeft.toString())) * // add event block timestamp + secsLeft
+        1000, // convert to ms
+    )
+    result = {
+      swapType,
+      settleableAtTimestamp,
+      secondsRemaining: Math.ceil(
+        (Date.now() - settleableAtTimestamp.getTime()) / 1000,
+      ),
```

Comment:
Note that `Date.now()` will always be head of `block.timestamp` (latest block's timestamp). Ideally we should be checking the remaining time against `(await ethers.provider.getBlock("latest")).timestamp` or read from the multicall contract https://github.com/makerdao/multicall/blob/master/src/Multicall.sol#L33-L35